jdk/test/sun/nio/cs/TestCharsetMapping.java

```diff
@@ -0,0 +1,640 @@
+        clearTests();
+        if (shiftHackDBCS) {
+            testFor(2).put(new byte[] { 0x0e });
+        }
+        csinfo.mappings.forEach(e -> {
+                if (e.cp2 != 0)
+                    return;          // skip composite (base+cc) for now
+                byte[] bs = e.bs;
+                char[] cc = Character.toChars(e.cp);
+                testFor(bs.length).put(bs, cc);
+            });
+        if (shiftHackDBCS) {
+            testFor(2).put(new byte[] { 0x0f });
+        }
+        for (int i = 0; i < maxBytesPerChar; i++) {
+            rv &= tests[i].run(DECODE);
+        }
+
+        // (3) ENCODE:
+        clearTests();
+        if (shiftHackDBCS) {
+            testFor(2).put(new byte[] { 0x0e });
+        }
+        csinfo.mappings.forEach(e -> {
+                if (e.cp2 != 0)
+                    return;          // skip composite (base+cc) for now
+                if (csinfo.nr != null && csinfo.nr.containsKey(e.bb))
+                    return;          // non-roundtrip b2c
+                if (csinfo.c2b != null && csinfo.c2b.containsKey(e.cp))
+                    return;          // c2b only mapping
+                byte[] bs = e.bs;
+                char[] cc = Character.toChars(e.cp);
+                testFor(bs.length).put(bs, cc);
+            });
+        if (csinfo.c2b != null)
+            csinfo.c2b.values().forEach(e -> {
+                    byte[] bs = e.bs;
+                    char[] cc = Character.toChars(e.cp);
+                    testFor(bs.length).put(bs, cc);
+                });
+        if (shiftHackDBCS) {
+            testFor(2).put(new byte[] { 0x0f });
+        }
+        for (int i = 0; i < maxBytesPerChar; i++) {
+            rv &= tests[i].run(ENCODE);
+        }
+        return rv;
+    }
+
+    private static class Entry {
+        byte[] bs;   // byte sequence reps
+        int cp;      // Unicode codepoint
+        int cp2;     // CC of composite
+        long bb;     // bs in "long" form for nr lookup;
+    }
+
+    private final static int  UNMAPPABLE = 0xFFFD;
+    private static final Pattern ptn = Pattern.compile("(?:0x)?(\\p{XDigit}++)\\s++(?:U\\+|0x)?(\\p{XDigit}++)(?:\\s++#.*)?");
+    private static final int G_BS  = 1;
+    private static final int G_CP  = 2;
+    private static final int G_CP2 = 3;
+
+    private static class CharsetInfo {
+        Charset  cs;
+        String   pkgName;
+        String   clzName;
+        String   csName;
+        String   hisName;
+        String   type;
+        boolean  isInternal;
+
+        // mapping entries
+        List<Entry> mappings;
+        Map<Long, Entry> nr;       // bytes -> entry
+        Map<Integer, Entry> c2b;   // cp -> entry
+
+        CharsetInfo(String csName, String clzName) {
+            this.csName = csName;
+            this.clzName = clzName;
+            if (csName.endsWith("_Solaris") ||
+                csName.endsWith("_MS5022X") ||
+                csName.endsWith("_MS932")) {
```

Comment:
I would probably move this to the `charsets()` parsing function. To have all the code that deals with different data formats between 8 and later versions in one place.

Also, why the suffix solution? Easier to understand would be just listing the affected 5 charsets by full name, that way one can directly compare with later versions of the charsets file.

Reply:
I think just laziness :) I noticed there were two common suffixes in use so could cut it down to three checks

I agree that making it explicit is clearer though and also avoids accidentally picking up any later additions with the same suffixes (though I think that's very unlikely to happen)